### DISCUSSION TITLE
How index RNA@data in Seurat V5?

### DISCUSSION THREAD

**Discussion by RobinPranter** [satijalab/seurat#8328] [Q&A] [answered]
Hi,

In previous versions of Seurat I used `in_SO@assays$RNA@data[in_sig,]` to access the data for a subset of the features (`in_SO `is my seurat object and `in_sig` containes a list of gene names eg `in_sig <- c("gene1", "gene2")`). In V5 this no longer works. 

Here are some outputs that might help:
`> in_SO@assays$RNA@data`
`Error: no slot of name "data" for this object of class "Assay5"`

`> in_SO@assays$RNA`
`Assay (v5) data with 25465 features for 3176 cells`
`First 10 features:`
` LOC114595581, LOC114595943, LOC114596289, LOC114585404, LOC114597600, LOC114600179, CWF19L1, LOC114604713, TUBGCP2, ZNF511`
`Layers:` 
`   counts`

`> in_SO@assays$RNA@layers$counts[in_sig,]`
`Error in .subscript.2ary(x, i, , drop = TRUE) : subscript out of bounds`

**Comment by samuel-marsh** [comment, marked as answer]
Hi,

Yes this is to be expected given the new structure of Seurat V5 objects (https://satijalab.org/seurat/articles/essential_commands).  This example also shows the benefit of using Seurat-based accessor functions as they will function across versions.

In this case what you are looking for is `FetchData`.

```
norm_data_subset <- FetchData(object = obj, vars = c("gene1", "gene2", etc), layer = "data")

# or as sparse matrix
norm_data_subset_mat <- as.sparse(FetchData(object = obj, vars = c("gene1", "gene2", etc), layer = "data"))
```

Best,
Sam

**Reply by RobinPranter** [reply]
Thank you verry much for your quick reply!

However this still does not seem to do the trick: 
`> FetchData(object = in_SO, vars = in_sig, layer = "data")`
`Warning: No layers found matching search pattern provided`
`Error in FetchData.Assay5(object = object[[DefaultAssay(object = object)]],  : `
`  layer "data" is not found in the object`

I guess it comes down to me not reaslly understanding where the data that was stored under RNA@data in V4 is stored in V5.

Perhaps this helps:
`> Assays(in_SO)`
`[1] "RNA"     "SCT_Reg"`
`> Layers(in_SO)`
`[1] "counts"`

**Reply by samuel-marsh** [reply]
Hi,

So the issue seems to be that the data hasn't been normalized.  What it is showing is that only a "counts" layer is present.  Therefore trying to extract normalized data (that is what the layer "data" is trying to find) will not work.  If you run `NormalizeData` and then run `FetchData` it should work just fine.

Best,
Sam

**Reply by RobinPranter** [reply]
Hi Samuel,

Thanks again for all your help. It turns out that the Layer "data" had moved from the Assay "RNA" to the Assay "SCT_Reg" which I generate using SCTransform(). 

`DefaultAssay(in_SO) <- "RNA"`
`Layers(in_SO)`
`[1] "counts"`

`> DefaultAssay(in_SO) <- "SCT_Reg"`
`> Layers(in_SO)`
`[1] "counts"     "data"       "scale.data"`

This makes sense to me, but I am not completely sure if this was the case in V4 or not since I changed version and normalization method somewhat around the same time. I think I changed method earlier though so I suspect it was not he case in V4?

Again, thanks a lot for your help!

Best,
Robin

**Comment by yingyonghui** [comment]
actually, `norm_data_subset <- FetchData(object = obj, vars = c("gene1", "gene2", etc), layer = "data")`
only get the matrix of selected genes.

How can I get all gene?
I run `norm_data_subset <- FetchData(object = obj, layer = "data")`
only to get the error "argument "vars" is missing, with no default"

**Reply by samuel-marsh** [reply]
Hi,
You can use `LayerData` to get all the data.

Best,
Sam
